### DISCUSSION TITLE
Using hash_routes and Zeitwerk

### DISCUSSION THREAD

**Discussion by Quintasan** [jeremyevans/roda#261] [General]
This is probably going to end up me understanding something incorrectly but please bear with me.

I was looking at using Zeitwerk with Roda for code reloading and I stumbled upon a problem:

1. `hash_routes` requires you to [reopen the main class for your application](https://github.com/jeremyevans/roda-sequel-stack/blob/master/routes/prefix1.rb)
2. Zeitwerk [expects filenames to reflect the classes ](https://github.com/fxn/zeitwerk#the-idea-file-paths-match-constant-paths) defined by them 

I had gotten around this by doing something I would call ["unsavoury hackery"](https://github.com/Quintasan/wireguard-dashboard-api/blob/master/config/environment.rb#L16) which would be to `load` the files under `routes/**/*.rb` and reload them using `listen` gem.

This brings me to the following question: would it be possible to respect Zeitgeist's file naming conventions and leverage `hash_routes` plugin at the same time?

**Comment by jeremyevans** [comment]
I'm not sure.  Because Zeitwerk has this limitation, I don't use it. I use https://github.com/jeremyevans/rack-unreloader for reloading, which can also reload individual route files instead of reloading the entire application if something changes (on applications of significant size, this is much faster).

**Reply by jeremyevans** [reply]
If you want an example of how to use rack-unreloader in a roda application, https://github.com/jeremyevans/roda-sequel-stack uses rack-unreloader to reload the roda app, roda route files, and sequel model files, so that modifications to one file only require reloading that file and not the entire application.

**Comment by janko** [comment]
I think this can work well if you configure Zeitwerk to [ignore](https://github.com/fxn/zeitwerk#ignoring-parts-of-the-project) the directory that don't follow Zeitwerk's conventions (in this case `routes/*.rb`), and load those files manually using [Zeitwerk callbacks](https://github.com/fxn/zeitwerk#callbacks).

I was able to get reloading to work in the following example project:

```rb
# config.ru
require_relative "config/boot"

run -> (env) { App.call(env) } # allows App to autoload
```
```rb
# config/boot.rb
require "bundler/setup"
require "zeitwerk"
require "listen"
require "pathname"

root = Pathname("..").expand_path(__dir__)

loader = Zeitwerk::Loader.new
loader.push_dir root

# load routes/*.rb files manually whenever App is loaded
loader.ignore root.join("routes")
loader.on_load("App") do
  root.glob("routes/*.rb") { |file| load(file) }
end

loader.enable_reloading if ENV["RACK_ENV"] == "development"
loader.setup

listener = Listen.to(root) { loader.reload }
listener.start
```
```rb
# app.rb
require "roda"

class App < Roda
  plugin :hash_routes

  route do |r|
    r.hash_branches
  end
end
```
```rb
# routes/articles.rb
class App
  hash_branch("articles") do |r|
    r.root do
      "Articles"
    end
  end
end
```

I needed to use `Kernel#load` instead of `Kernel#require`, as `#require` skips loading if the file has already been loaded. An alternative would be to first remove the files from `$LOADED_FEATURES`, and then call `#require`.

**Reply by HoneyryderChuck** [reply]
That's pretty impressive, never went that far! But the whole of the routes dir has to be reloaded, right?

**Reply by janko** [reply]
Yeah, all route files are loaded whenever `App` loads. I'm guessing it would be better to load them in `app.rb` instead of `on_load`, as Xavier showed in https://github.com/jeremyevans/roda-sequel-stack/issues/21#issuecomment-894451999. But the dependencies of the route files (models, services etc.) are still autoloaded lazily, so I was under the impression that route files themselves would load fast.

The approach that I would personally use, with which routes could be autoloaded lazily, would be to run sub-apps:

```rb
# app.rb
class App < Roda
  route do |r|
    r.on "albums" do |r|
      r.run Routes::Albums
    end
    # ...
  end
end
```
```rb
# routes/albums.rb
class Routes::Albums < App
  route do |r|
    # ...
  end
end
```

This is essentially what [multi_run](https://roda.jeremyevans.net/rdoc/classes/Roda/RodaPlugins/MultiRun.html) plugin does, except that because Roda sub-apps are only referenced inside the `route` block, they are loaded lazily. We used this approach with Cuba at a previous company, though not for reloading purposes (I think we used shotgun/rerun).

**Comment by HoneyryderChuck** [comment]
I also asked this question before, see https://github.com/jeremyevans/roda-sequel-stack/issues/21

**Reply by topofocus** [reply]
Interesting discussion.

During my scanning of working roda concepts  I  found a similar approach

The app is called 'ReviewAnything' 

It follows that pattern. But based the routing-classes from roda instead of app.

```ruby
class App < Roda
  route do |r|
     r.root { view "home" }
     r.on( 'users' ){ r.run UserController }
    (...)
  end
end

class UserController < Roda
  route do | r | 
      r.is  do
       r.post do 
     (..)
       end
     end
   (...)
  end
```

As I understand, inheriting the routes (controllers) from App includes all plugins. 
The alternative implementation offers individual plugin-configurations.

Am I right?

**Reply by jeremyevans** [reply]
You are correct.  Note that if you want to use this approach, the `multi_run` plugin may be useful to you.  The downside to this approach is it is probably slower than the `hash_routes` approach, though I doubt it matters much in a typical application.
